command/operator_diagnose.go

```diff
@@ -173,119 +186,270 @@ func (c *OperatorDiagnoseCommand) offlineDiagnostics(ctx context.Context) error
-		disableClustering := config.HAStorage.DisableClustering
-		infoKeys := make([]string, 0, 10)
-		info := make(map[string]string)
-		status, lns, _, errMsg := server.InitListeners(config, disableClustering, &infoKeys, &info)
+	var metricSink *metricsutil.ClusterMetricSink
+	var metricsHelper *metricsutil.MetricsHelper
 
-		if status != 0 {
-			return errMsg
-		}
+	var backend *physical.Backend
+	diagnose.Test(ctx, "storage", func(ctx context.Context) error {
+		diagnose.Test(ctx, "create-storage-backend", func(ctx context.Context) error {
 
-		// Make sure we close all listeners from this point on
-		listenerCloseFunc := func() {
-			for _, ln := range lns {
-				ln.Listener.Close()
+			b, err := server.setupStorage(config)
+			if err != nil {
+				return err
 			}
+			backend = &b
+			return nil
+		})
+
+		if config.Storage == nil {
+			return fmt.Errorf("no storage stanza found in config")
 		}
 
-		defer c.cleanupGuard.Do(listenerCloseFunc)
+		if config.Storage != nil && config.Storage.Type == storageTypeConsul {
+			diagnose.Test(ctx, "test-storage-tls-consul", func(ctx context.Context) error {
+				err = physconsul.SetupSecureTLS(api.DefaultConfig(), config.Storage.Config, server.logger, true)
+				if err != nil {
+					return err
+				}
+				return nil
+			})
 
-		sanitizedListeners := make([]listenerutil.Listener, 0, len(config.Listeners))
-		for _, ln := range lns {
-			if ln.Config.TLSDisable {
-				diagnose.Warn(ctx, "TLS is disabled in a Listener config stanza.")
-				continue
-			}
-			if ln.Config.TLSDisableClientCerts {
-				diagnose.Warn(ctx, "TLS for a listener is turned on without requiring client certs.")
-			}
+			diagnose.Test(ctx, "test-consul-direct-access-storage", func(ctx context.Context) error {
+				dirAccess := diagnose.ConsulDirectAccess(config.Storage.Config)
+				if dirAccess != "" {
+					diagnose.Warn(ctx, dirAccess)
+				}
+				return nil
+			})
+		}
+
+		// Attempt to use storage backend
+		if !c.skipEndEnd {
+			diagnose.Test(ctx, "test-access-storage", diagnose.WithTimeout(30*time.Second, func(ctx context.Context) error {
+				// TODO: A static file in storage that probably won't cause a collision seems low-risk to write to for now.
+				// Should we make this a proper uuid?
+
+				veryRandomUuid := "diagnose-secret-uuid-1234"
+				err := diagnose.EndToEndLatencyCheckWrite(ctx, veryRandomUuid, *backend)
+				if err != nil && strings.Contains(err.Error(), diagnose.LatencyWarning) {
+					diagnose.Warn(ctx, err.Error())
+				} else if err != nil {
+					return err
+				}
+				err = diagnose.EndToEndLatencyCheckRead(ctx, veryRandomUuid, *backend)
+				if err != nil && strings.Contains(err.Error(), diagnose.LatencyWarning) {
+					diagnose.Warn(ctx, err.Error())
+				} else if err != nil {
+					return err
+				}
+				err = diagnose.EndToEndLatencyCheckDelete(ctx, veryRandomUuid, *backend)
+				if err != nil && strings.Contains(err.Error(), diagnose.LatencyWarning) {
```

Comment:
It's pretty likely if latency is high for one of these, it's high for all of them.  Would it be possible to create a custom error with the latency as a field, then just report once with the highest observed latency?

Reply:
Sure!